File: modules/stockity-websocket/ios/StockityWebSocketModule.swift
import ExpoModulesCore

public class StockityWebSocketModule: Module {
  private var sockets: [String: URLSessionWebSocketTask] = [:]
  private var sessions: [String: URLSession] = [:]
  private var delegates: [String: StockityWebSocketDelegate] = [:]

  public func definition() -> ModuleDefinition {
    Name("StockityWebSocket")

    Events("open", "message", "error", "close")

    AsyncFunction("connect") { (url: String, headers: [String: String]) -> String in
      guard let wsUrl = URL(string: url) else {
        throw NSError(domain: "StockityWebSocket", code: 0, userInfo: [
          NSLocalizedDescriptionKey: "Invalid URL"
        ])
      }
      let socketId = UUID().uuidString
      var request = URLRequest(url: wsUrl)
      headers.forEach { key, value in
        request.setValue(value, forHTTPHeaderField: key)
      }
      let delegate = StockityWebSocketDelegate(socketId: socketId, module: self)
      let session = URLSession(configuration: .default, delegate: delegate, delegateQueue: nil)
      let task = session.webSocketTask(with: request)
      sockets[socketId] = task
      sessions[socketId] = session
      delegates[socketId] = delegate
      task.resume()
      listen(socketId)
      return socketId
    }

    AsyncFunction("send") { (socketId: String, message: String) -> Bool in
      guard let task = sockets[socketId] else { return false }
      task.send(.string(message)) { error in
        if let error = error {
          self.sendEvent("error", ["id": socketId, "message": error.localizedDescription])
        }
      }
      return true
    }

    AsyncFunction("close") { (socketId: String, code: Int?, reason: String?) in
      guard let task = sockets.removeValue(forKey: socketId) else { return }
      let closeCode = URLSessionWebSocketTask.CloseCode(rawValue: code ?? 1000) ?? .normalClosure
      let reasonData = reason?.data(using: .utf8)
      task.cancel(with: closeCode, reason: reasonData)
      sessions[socketId]?.invalidateAndCancel()
      sessions.removeValue(forKey: socketId)
      delegates.removeValue(forKey: socketId)
    }
  }

  private func listen(_ socketId: String) {
    guard let task = sockets[socketId] else { return }
    task.receive { result in
      switch result {
      case .failure(let error):
        self.sendEvent("error", ["id": socketId, "message": error.localizedDescription])
      case .success(let message):
        switch message {
        case .string(let text):
          self.sendEvent("message", ["id": socketId, "data": text])
        case .data(let data):
          let text = String(data: data, encoding: .utf8) ?? ""
          self.sendEvent("message", ["id": socketId, "data": text])
        @unknown default:
          break
        }
      }
      self.listen(socketId)
    }
  }
}

public class StockityWebSocketDelegate: NSObject, URLSessionWebSocketDelegate, URLSessionTaskDelegate {
  private let socketId: String
  private weak var module: StockityWebSocketModule?

  init(socketId: String, module: StockityWebSocketModule) {
    self.socketId = socketId
    self.module = module
  }

  public func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask, didOpenWithProtocol protocol: String?) {
    var payload: [String: Any] = ["id": socketId]
    if let response = webSocketTask.response as? HTTPURLResponse {
      payload["responseCode"] = response.statusCode
      payload["responseHeaders"] = response.allHeaderFields.description
    }
    module?.sendEvent("open", payload)
  }

  public func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask, didCloseWith closeCode: URLSessionWebSocketTask.CloseCode, reason: Data?) {
    let reasonText = reason.flatMap { String(data: $0, encoding: .utf8) } ?? ""
    module?.sendEvent("close", ["id": socketId, "code": closeCode.rawValue, "reason": reasonText])
  }

  public func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
    if let error = error {
      var payload: [String: Any] = [
        "id": socketId,
        "message": error.localizedDescription
      ]
      if let response = task.response as? HTTPURLResponse {
        payload["responseCode"] = response.statusCode
        payload["responseHeaders"] = response.allHeaderFields.description
      }
      module?.sendEvent("error", payload)
    }
  }
}

File: modules/stockity-websocket/android/src/main/java/expo/modules/stockitywebsocket/StockityWebSocketModule.kt
package expo.modules.stockitywebsocket

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

class StockityWebSocketModule : Module() {
  private val client = OkHttpClient.Builder().build()
  private val sockets = ConcurrentHashMap<String, WebSocket>()

  override fun definition() = ModuleDefinition {
    Name("StockityWebSocket")

    Events("open", "message", "error", "close")

    AsyncFunction("connect") { url: String, headers: Map<String, String> ->
      val socketId = UUID.randomUUID().toString()
      val requestBuilder = Request.Builder().url(url)
      headers.forEach { (key, value) ->
        requestBuilder.addHeader(key, value)
      }
      val request = requestBuilder.build()
      val socket = client.newWebSocket(
        request,
        object : WebSocketListener() {
          override fun onOpen(webSocket: WebSocket, response: Response) {
            sendEvent(
              "open",
              mapOf(
                "id" to socketId,
                "responseCode" to response.code,
                "responseMessage" to response.message,
                "responseHeaders" to response.headers.toString()
              )
            )
          }

          override fun onMessage(webSocket: WebSocket, text: String) {
            sendEvent("message", mapOf("id" to socketId, "data" to text))
          }

          override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
            sendEvent(
              "error",
              mapOf(
                "id" to socketId,
                "message" to (t.message ?: "WebSocket failure"),
                "responseCode" to response?.code,
                "responseMessage" to response?.message,
                "responseHeaders" to response?.headers?.toString()
              )
            )
            sockets.remove(socketId)
          }

          override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
            sendEvent("close", mapOf("id" to socketId, "code" to code, "reason" to reason))
            sockets.remove(socketId)
          }
        }
      )
      sockets[socketId] = socket
      socketId
    }

    AsyncFunction("send") { socketId: String, message: String ->
      val socket = sockets[socketId]
      socket?.send(message) ?: false
    }

    AsyncFunction("close") { socketId: String, code: Int?, reason: String? ->
      val socket = sockets.remove(socketId) ?: return@AsyncFunction
      val closeCode = code ?: 1000
      socket.close(closeCode, reason ?: "")
    }
  }
}
